File: solutions/3.py
# sliding window problem, но есть два варианта. Закидывать посещенные в set и добавлять когда встретили, но тогда чтобы понять когда 
# мы последний раз встречали этот элемент мы должны его еще найти. Так что лучше использовать словарь и туда запихивать когда последний
# раз мы видели этот элемент чтобы с него начать наш window
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0

        result = 1
        seen_chars = {}
        start, i = 0, 0
        while i < len(s):
            print("------")
            print(start, i, result)
            print(s[start:i])
            if s[i] not in seen_chars:
                seen_chars[s[i]] = i
            elif s[i] in seen_chars:
                if start <= seen_chars[s[i]] <= i:
                    start = seen_chars[s[i]] + 1
                seen_chars[s[i]] = i
            result = max(result, i - start + 1)
            i += 1
        return result

s = "ab!bb"
solution = Solution()
print(solution.lengthOfLongestSubstring(s))

File: solutions/3.go
package main


func lengthOfLongestSubstring(s string) int {
    if len(s) == 0 {
        return 0
    }
    result := 1
    currChars := map[byte]int{}
    l, r := 0, 0
    for r < len(s) {
        val, ok := currChars[s[r]]
        if !ok {
            currChars[s[r]] = r
        } else {
            if l <= val && val <= r {
                l = val + 1
            }
            currChars[s[r]] = r
        }
        result = max(result, r - l + 1)
        r++
    }
    return max(result, r - l)
}